### PR TITLE
Add get update delete account functionality

### PR DESCRIPTION
**Pull request by ekoval**
Add Get, Update, Delete account functionality
Unittests code is re-organized and made more flexible.
Related to: https://github.com/yola/production/issues/2116

### CODE REVIEW

File: tests/test_integration_account.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
+import base64
+import uuid
+
+from demands import HTTPServiceError
+
+from base import BaseTestCase, BaseObjectDoesntExistTestCase
```
Comment (snitch) [inline tests/test_integration_account.py]: I101 Imported names are in the wrong order. Should be BaseObjectDoesntExistTestCase, BaseTestCase

Comment (snitch) [inline tests/test_integration_account.py]: I101 Imported names are in the wrong order. Should be BaseObjectDoesntExistTestCase, BaseTestCase

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

File: sitewit/models.py [hunk 2]
```diff
@@ -52,3 +54,62 @@ def create(cls, user, site, url, user_token=None):
...
+        Args:
+            account_token (str): account token.
+
+        Returns:
+            Instance of Account class.
+
+        Raises:
+            demands.HTTPServiceError: if any error happened on HTTP level.
+        """
+        result = cls.get_service().delete_account(account_token)
```
Comment (sgalaburda) [inline sitewit/models.py]: Do we really need to return any data from the delete call?

File: sitewit/services.py [hunk 3]
```diff
@@ -65,3 +65,52 @@ def create_account(self, site_id, url, user_name, user_email,
...
+        Args:
+            account_token (str): account token.
+
+        Returns:
+            account json:
+            https://sandboxpapi.sitewit.com/Help/Api/GET-api-Account
+        """
+        return self.get(
+            '/api/account/',
+            headers=self._get_auth_header(account_token)).json()
```
Comment (sgalaburda) [inline sitewit/services.py]: There is a common pattern `self.METHOD('/api/account/', headers=self._get_auth_header(account_token) ).json()` we can move it to a separate method to reduce code duplication.

File: tests/accounts/base.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+from demands import HTTPServiceError
+from mock import Mock
+from unittest2 import TestCase, skip
+from yoconfig import configure
+from yoconfigurator.base import read_config
+
+from sitewit.services import SitewitService
+
+
+class BaseTestCase(TestCase):
```
Comment (sgalaburda) [inline tests/accounts/base.py]: @blaix said that the code smells when `Base` is used in the name of the base class.

File: sitewit/models.py (hunk 2 above)
Comment (ekoval) [inline sitewit/models.py]: API does this. Also, we had a discussion with @RayeN (Voffka) before and decided that alsays returning modified instance is a way-to-go.

Comment (ekoval) [inline sitewit/models.py]: API does this. Also, we had a discussion with @RayeN (Voffka) before and decided that alsays returning modified instance is a way-to-go.

File: tests/accounts/base.py (hunk 4 above)
Comment (ekoval) [inline tests/accounts/base.py]: Ahh, maybe you remember why he said so? :)

Comment (ekoval) [inline tests/accounts/base.py]: Ahh, maybe you remember why he said so? :)

File: sitewit/services.py (hunk 3 above)
Comment (ekoval) [inline sitewit/services.py]: That's a good idea, I was thinking about this a bit, but then decided that the profit in this specific case isn't worth efforts. Also, please note that we have different headers for two different auth cases.

File: tests/accounts/base.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+        if data is not None:
+            demands_mock.assert_called_once_with(
+                '/api/account/', data, headers=headers)
+        else:
+            demands_mock.assert_called_once_with(
+                '/api/account/', headers=headers)
+
+
+    def _mock_response(self, requests_mock, response):
```
Comment (snitch) [inline tests/accounts/base.py]: E303 too many blank lines (2)

File: tests/accounts/test_models.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+class TestModelsGetAccount(BaseTestCase):
+    @patch.object(sitewit.models.SitewitService, 'get')
+    def setUp(self, get_mock):
+        self.get_mock = get_mock
+        self._mock_response(get_mock, self.response_brief)
+
+        self.result = Account.get(self.token)
+
+    def test_demands_get_is_called(self):
+        partner_id = self.config.common.sitewit['affiliate_id']
```
Comment (snitch) [inline tests/accounts/test_models.py]: F841 local variable 'partner_id' is assigned to but never used

File: tests/accounts/test_models.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        partner_token = self.config.common.sitewit['affiliate_token']
```
Comment (snitch) [inline tests/accounts/test_models.py]: F841 local variable 'partner_token' is assigned to but never used

File: tests/accounts/test_models.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+    def test_demands_put_is_called(self):
+        put_data = {
+            'url': self.url,
+            'currency': self.currency,
+            'countryCode': self.country_code,
+            'timeZone': 'GMT Standard Time',
+        }
+
+        partner_id = self.config.common.sitewit['affiliate_id']
```
Comment (snitch) [inline tests/accounts/test_models.py]: F841 local variable 'partner_id' is assigned to but never used

File: tests/accounts/test_models.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        partner_token = self.config.common.sitewit['affiliate_token']
```
Comment (snitch) [inline tests/accounts/test_models.py]: F841 local variable 'partner_token' is assigned to but never used

File: tests/accounts/test_models.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+class TestModelsDeleteAccount(BaseTestCase):
+    @patch.object(sitewit.models.SitewitService, 'delete')
+    def setUp(self, delete_mock):
+        self.delete_mock = delete_mock
+        self._mock_response(delete_mock, self.response_brief)
+
+        self.account = Account.delete(self.token)
+
+    def test_demands_delete_is_called(self):
+        partner_id = self.config.common.sitewit['affiliate_id']
```
Comment (snitch) [inline tests/accounts/test_models.py]: F841 local variable 'partner_id' is assigned to but never used

File: tests/accounts/test_models.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+        partner_token = self.config.common.sitewit['affiliate_token']
```
Comment (snitch) [inline tests/accounts/test_models.py]: F841 local variable 'partner_token' is assigned to but never used

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

File: tests/accounts/base.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+        self.assertEqual(account.site_id, self.site_id)
+        self.assertEqual(account.currency, self.currency)
+        self.assertEqual(account.country_code, self.country_code)
+        self.assertEqual(account.status, 'Active')
+        self.assertEqual(account.token, self.token)
+
+        if check_user_info:
+            self.assertEqual(account.user.name, self.user_name)
+            self.assertEqual(account.user.email, self.user_email)
+            self.assertEqual(account.user.token, self.user_token)
```
Comment (RayeN) [inline tests/accounts/base.py]: I would move it t different assertation

File: tests/accounts/base.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+class BaseObjectDoesntExistTestCase(BaseTestCase):
+    operation = None
+
+    def setUp(self):
+        self.service = SitewitService()
+
+    def test_exception_is_raised(self):
+        if self.operation is None:
+            return skip("BaseTest tests skipped")
```
Comment (RayeN) [inline tests/accounts/base.py]: double quotes

File: tests/accounts/base.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+        requests_mock.return_value = response_mock
+
+
+class BaseObjectDoesntExistTestCase(BaseTestCase):
+    operation = None
+
+    def setUp(self):
+        self.service = SitewitService()
+
+    def test_exception_is_raised(self):
```
Comment (RayeN) [inline tests/accounts/base.py]: probably it will be clearer to make it assertion?

File: tests/accounts/test_integration.py [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+
+        expected_error_details = {
+            u'ModelState': {
+                u'account.url': [u'Invalid Url']
+            },
+            u'Message': u'The request is invalid.'
+        }
+
+        self.assertEqual(exc.exception.response.status_code, 400)
+        self.assertEqual(exc.exception.details, expected_error_details)
```
Comment (RayeN) [inline tests/accounts/test_integration.py]: you already have a special helper for this test, nope?

File: tests/accounts/base.py [hunk 16]
```diff
@@ -0,0 +1,97 @@
+import base64
+import os
+import uuid
+
+from demands import HTTPServiceError
+from mock import Mock
+from unittest2 import TestCase, skip
```
Comment (snitch) [inline tests/accounts/base.py]: F401 'skip' imported but unused

File: tests/accounts/base.py [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+from yoconfig import configure
+from yoconfigurator.base import read_config
+
+from sitewit.services import SitewitService
```
Comment (snitch) [inline tests/accounts/base.py]: F401 'SitewitService' imported but unused

File: tests/accounts/base.py [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+            demands_mock.assert_called_once_with(
+                '/api/account/', headers=headers)
+
+    def _mock_response(self, requests_mock, response):
+        response_mock = Mock()
+        response_mock.json = Mock(return_value=response)
+        requests_mock.return_value = response_mock
+
+    def assertHTTPErrorIsRaised(
+        self, method, params, expected_code, expected_details):
```
Comment (snitch) [inline tests/accounts/base.py]: E125 continuation line with same indent as next logical line

File: tests/accounts/test_integration.py [hunk 19]
```diff
@@ -0,0 +1,193 @@
+import base64
```
Comment (snitch) [inline tests/accounts/test_integration.py]: F401 'base64' imported but unused

File: tests/accounts/test_integration.py [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+import uuid
+
+from demands import HTTPServiceError
```
Comment (snitch) [inline tests/accounts/test_integration.py]: F401 'HTTPServiceError' imported but unused

File: tests/accounts/test_integration.py [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+        expected_error_details = {
+            u'ModelState': {
+                u'account.url': [u'Missing url parameter']
+            },
+            u'Message': u'The request is invalid.'
+        }
+
+        self.assertHTTPErrorIsRaised(
+            SitewitService().create_account, (
+                '', '', self.user_name,self.user_email, self.currency,
```
Comment (snitch) [inline tests/accounts/test_integration.py]: E231 missing whitespace after ','

File: tests/accounts/test_integration.py [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+                u'account.url': [u'Invalid Url']
+            },
+            u'Message': u'The request is invalid.'
+        }
+
+        self.assertHTTPErrorIsRaised(
+            self.service.update_account, (self.token, 'aa', 'GB', 'GBP'),
+            400, expected_error_details)
+
+class TestUpdateAccountDoesNotExist(BaseTestCase):
```
Comment (snitch) [inline tests/accounts/test_integration.py]: E302 expected 2 blank lines, found 1

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

File: tests/accounts/base.py [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+import uuid
+
+from demands import HTTPServiceError
+from mock import Mock
+from unittest2 import TestCase
+from yoconfig import configure
+from yoconfigurator.base import read_config
+
+
+class BaseTestCase(TestCase):
```
Comment (blaix) [inline tests/accounts/base.py]: @sgalaburda commented on a diff that is now outdated saying that I consider classes with `Base` in the name to be a code smell. @ekoval asked why and I never replied.

Now I am replying. Keep in mind I haven't looked at this code, I'm just explaining my opinion:

It is too generic. Giving something the name "base" makes it easy to become a dumping ground for any functionality you might want somewhere else. So I think it's usually a sign that a more expressive name is needed, or that it needs to be split into smaller, more cohesive pieces that are easily named for what they are.

**Comment by RayeN** [comment]
:+1:

File: tests/accounts/base.py (hunk 23 above)
Comment (blaix) [inline tests/accounts/base.py]: https://github.com/garybernhardt/base (a joke)

**Pull request merged**